File: cpp_pkg/src/class_client_add_two_ints.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"


class AddTwoIntsClient: public rclcpp::Node{

    private:

        std::thread thread_;

    public:

        AddTwoIntsClient():Node("client_add_two_ints"){
            thread_ = std::thread(std::bind(&AddTwoIntsClient::call_server, this, 13, 22));
        }

        void call_server(int a, int b){
            auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            while (!client->wait_for_service(std::chrono::seconds(1)))
                RCLCPP_WARN(this->get_logger(), "Waiting for service");

            request->a = a;
            request->b = b;
            
            auto future = client->async_send_request(request);

            try 
            {
                auto response = future.get();
                RCLCPP_INFO(this->get_logger(), "%d + %d = %d", a, b, (int)response->sum);
            }
            
            catch(std::exception &ex){
                RCLCPP_ERROR(this->get_logger(), "ERROR: Service call failed");
            }
        }

};

int main(int argc, char **argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClient>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;


}

File: python_packages/python_pkg/python_pkg/class_client_add_two_ints.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

class ClientAddTwoInts(Node):
    def __init__(self):
        super().__init__("client_add_two_ints")

    def call_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        request = AddTwoInts.Request()

        request.a = a
        request.b = b

        while not client.wait_for_service(1):
            self.get_logger().warn("Waiting for server Add Two Ints")

        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_call, a=a, b=b))

    def callback_call(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f"{a} + {b} = {response.sum}")

        except Exception as e:
            self.get_logger().error("Service call error %r" %(e,))


def main(args = None):
    rclpy.init(args=args)
    node = ClientAddTwoInts()

    i1 = 43
    i2 = 54

    for i in range (0, 10):
        node.call_server(i1 + i, i2 + i)

    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
